### PR TITLE
abci: Remove SetOption #5447 #9091

### PR DESCRIPTION
**Pull request by samricotta**
Moving the changes from issue #5447 into main. This PR removes theResponse/Request SetOption from ABCI replicated from https://github.com/tendermint/tendermint/pull/5447

Note: This is part of the ABCI++ Prepare/Process Proposal Release Plan #9091

### CODE REVIEW

File: docs/app-dev/abci-cli.md [hunk 1]
```diff
@@ -285,9 +285,6 @@ In another window, start the `abci-cli console`:
 
 ```sh
 > set_option serial on
```
Comment (cmwaters) [inline docs/app-dev/abci-cli.md:287]: I think you need to remove this line as well. It looks like there's a few places further up where we mention `setOption`. I don't think this is necessary because in https://github.com/tendermint/tendermint/pull/9143 I am removing the counter app entirely

File: proto/tendermint/abci/types.proto [hunk 2]
```diff
@@ -37,6 +36,7 @@ message Request {
     RequestLoadSnapshotChunk  load_snapshot_chunk  = 14;
     RequestApplySnapshotChunk apply_snapshot_chunk = 15;
   }
+  Reserve = 4;
```
Comment (cmwaters) [inline proto/tendermint/abci/types.proto:39]:
```suggestion
  reserved 4;
```

File: proto/tendermint/abci/types.proto [hunk 3]
```diff
@@ -146,6 +139,7 @@ message Response {
     ResponseLoadSnapshotChunk  load_snapshot_chunk  = 15;
     ResponseApplySnapshotChunk apply_snapshot_chunk = 16;
   }
+  Reserve = 5;
```
Comment (cmwaters) [inline proto/tendermint/abci/types.proto:142]:
```suggestion
  reserved 5;  // was setOption
```

**Comment by cmwaters** [comment]
May be best to base this against `feature/abci++ppp` as this change is involved with abci++. We might also want to leave it until most of the ppp is done before merging. wdyt @sergio-mena?

**Comment by williambanfield** [comment]
It looks like this changes a lot of files not touched in #5447 , I think we should make sure to just update those files with this change since I don't see a strong reason to edit the workflows to make this work?

**Comment by sergio-mena** [comment]
> May be best to base this against `feature/abci++ppp` as this change is involved with abci++. We might also want to leave it until most of the ppp is done before merging. wdyt @sergio-mena?

Yeah, merging into `feature/abci++ppp` would be the best option at this time IMO.

File: abci/tests/test_cli/ex2.abci [hunk 4]
```diff
@@ -1,4 +1,3 @@
-set_option serial on
```
Comment (sergio-mena) [inline abci/tests/test_cli/ex2.abci:1]: Did you test this manually?

File: proto/tendermint/abci/types.proto [hunk 5]
```diff
@@ -37,6 +36,7 @@ message Request {
     RequestLoadSnapshotChunk  load_snapshot_chunk  = 14;
     RequestApplySnapshotChunk apply_snapshot_chunk = 15;
   }
+  reserved = 4;
```
Comment (sergio-mena) [inline proto/tendermint/abci/types.proto:39]: Nice reserving the value!
Not sure the '=' is valid syntax (as the linter points out)

File: proto/tendermint/abci/types.proto [hunk 6]
```diff
@@ -146,6 +139,7 @@ message Response {
     ResponseLoadSnapshotChunk  load_snapshot_chunk  = 15;
     ResponseApplySnapshotChunk apply_snapshot_chunk = 16;
   }
+  reserved = 5;  // was setOption
```
Comment (sergio-mena) [inline proto/tendermint/abci/types.proto:142]:
```suggestion
  reserved 5;  // was setOption
```

File: spec/core/data_structures.md [hunk 7]
```diff
@@ -130,7 +130,7 @@ the data in the current block, the previous block, and the results returned by t
...
-| EvidenceHash      | slice of bytes (`[]byte`) | MerkleRoot of the evidence of Byzantine behaviour included in this block.                                                                                                                                                                                                                                                                                                             | Must  be of length 32                                                                                                                                                                            |
```
Comment (sergio-mena) [inline spec/core/data_structures.md:133]: Heads-up: if you're changing this to appease the linter, it's not really necessary on ".md" files; only go files seem to be checked (known pain :-D)

**Comment by samricotta** [comment]
Currently blocked by ABCI++ work

File: CHANGELOG_PENDING.md [hunk 8]
```diff
@@ -13,14 +13,12 @@ Friendly reminder, we have a [bug bounty program](https://hackerone.com/tendermi
 - Apps
 
   - [abci/counter] \#6684 Delete counter example app
-  - [txResults] \#9175 Remove `gas_used` & `gas_wanted` from being merkelized in the lastresulthash in the header
-  - [abci] \#5783 Make length delimiter encoding consistent (`uint64`) between ABCI and P2P wire-level protocols
```
Comment (thanethomson) [inline CHANGELOG_PENDING.md:17]: Why are these changelog_pending entries deleted here?

File: abci/example/counter/counter.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
+package counter
```
Comment (thanethomson) [inline abci/example/counter/counter.go:1]: Wasn't this file supposed to have been removed in https://github.com/tendermint/tendermint/commit/ea271c534a6955b8abc9e2b5b62bd16a437a81a4?

File: abci/tests/test_app/app.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
+package main
```
Comment (thanethomson) [inline abci/tests/test_app/app.go:1]: This also looks like it was supposed to be removed in https://github.com/tendermint/tendermint/commit/ea271c534a6955b8abc9e2b5b62bd16a437a81a4 right?

File: proto/tendermint/abci/types.proto (hunk 5 above)
Comment (samricotta) [reply proto/tendermint/abci/types.proto]: The reserve should be formatted correctly now

File: docs/app-dev/abci-cli.md [hunk 11]
```diff
@@ -213,9 +214,137 @@ we do `deliver_tx "abc=efg"` it will store `(abc, efg)`.
 Similarly, you could put the commands in a file and run
 `abci-cli --verbose batch < myfile`.
 
+## Counter - Another Example
```
Comment (sergio-mena) [inline docs/app-dev/abci-cli.md:217]: I think we have removed Counter App in `main`, and since we have merge `main` into `feature/abci++ppp` it also applies to the latter

File: proto/tendermint/abci/types.proto [hunk 12]
```diff
@@ -128,12 +122,12 @@ message RequestApplySnapshotChunk {
 // Response types
 
 message Response {
+  reserved 5;
```
Comment (sergio-mena) [inline proto/tendermint/abci/types.proto:125]:
```suggestion
  reserved 5; // ResponseSetOption removed
```

File: proto/tendermint/abci/types.proto [hunk 13]
```diff
@@ -20,11 +20,11 @@ import "gogoproto/gogo.proto";
 // Request types
 
 message Request {
+  reserved 4;
```
Comment (sergio-mena) [inline proto/tendermint/abci/types.proto:23]:
```suggestion
  reserved 4; // RequestSetOption removed
```

File: tools/tm-signer-harness/README.md [hunk 14]
```diff
@@ -1,5 +0,0 @@
-# tm-signer-harness
-
-See the [`tm-signer-harness`
-documentation](https://tendermint.com/docs/tools/remote-signer-validation.html)
-for more details.
```
Comment (williambanfield) [inline tools/tm-signer-harness/README.md:5]: Not sure why this change was included here.

**Review by williambanfield — COMMENTED** [review]
Still seeing references to set option in a few places in the repo:

```
rg "set_option"
docs/app-dev/abci-cli.md
41:  set_option  Set an options on the application
265:We can toggle the value of `serial` using the `set_option` ABCI message.
287:> set_option serial on

abci/tests/test_cli/ex2.abci.out
1:> set_option serial on
```
